apps/server/src/utils/sheetsAuth.ts

```diff
@@ -0,0 +1,382 @@
+import { sheets, sheets_v4 } from '@googleapis/sheets';
+import { writeFile } from 'fs/promises';
+import { readFileSync } from 'fs';
+import { OAuth2Client } from 'google-auth-library';
+import http from 'http';
+import { DatabaseModel, LogOrigin } from 'ontime-types';
+import { join } from 'path';
+import { URL } from 'url';
+import { logger } from '../classes/Logger.js';
+import { DataProvider } from '../classes/data-provider/DataProvider.js';
+import { getAppDataPath } from '../setup.js';
+import { ensureDirectory } from './fileManagement.js';
+import { cellRequenstFromEvent, cellRequenstFromProjectData, getA1Notation } from './sheetUtils.js';
+import { parseExcel } from './parser.js';
+import { parseProject, parseRundown, parseUserFields } from './parserFunctions.js';
+import { ExcelImportMap } from 'ontime-utils';
+
+type ResponseOK = {
+  data: Partial<DatabaseModel>;
+};
+
+class Sheet {
+  private static client: null | OAuth2Client = null;
+  private readonly scope = 'https://www.googleapis.com/auth/spreadsheets';
+  private readonly sheetsFolder: string;
+  private readonly clientSecretFile: string;
+  private static clientSecret = null;
+  private static authUrl: null | string = null;
+  private authServerTimeout;
+
+  private readonly requiredClientKeys = [
+    'client_id',
+    'project_id',
+    'auth_uri',
+    'token_uri',
+    'auth_provider_x509_cert_url',
+    'client_secret',
+    'redirect_uris',
+  ];
+
+  constructor() {
+    const appDataPath = getAppDataPath();
+    if (appDataPath === '') {
+      throw new Error('Sheet: Could not resolve sheet folser');
+    }
+    this.sheetsFolder = join(appDataPath, 'sheets');
+    this.clientSecretFile = join(this.sheetsFolder, 'client_secret.json');
+    ensureDirectory(this.sheetsFolder);
+    try {
+      const secrets = JSON.parse(readFileSync(this.clientSecretFile, 'utf-8'));
+      const isKeyMissing = this.requiredClientKeys.some((key) => !(key in secrets['installed']));
```

Comment:
This does not work for me. When uploading the Client ID (actually, the file is called client_secret_... ; maybe we should call it that, too, at least somewhere to not confuse people who don't read the docs? ;-) ), I get the following error: `Internal Server Error, 'TypeError: Cannot use 'in' operator to search for 'client_id' in undefined"`.
My secret file generated by google looks like this:
```json
{
    "web": {
        "client_id": "redacted.apps.googleusercontent.com",
        "project_id": "redacted",
        "auth_uri": "https://accounts.google.com/o/oauth2/auth",
        "token_uri": "https://oauth2.googleapis.com/token",
        "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
        "client_secret": "redacted",
        "redirect_uris": [
            "http://localhost"
        ]
    }
}
```